rust/src/interpreter/utils.rs

```diff
@@ -0,0 +1,127 @@
+use std::cmp::min;
+
+use evmc_vm::{ExecutionMessage, MessageFlags, Revision, StatusCode};
+
+use crate::{interpreter::gas::consume_copy_cost, types::u256};
+
+#[inline(always)]
+pub fn get_mut_slice_within_bounds<T>(data: &[T], offset: u256, len: u64) -> &[T] {
+    if len == 0 {
+        return &[];
+    }
+    let (offset, offset_overflow) = offset.into_u64_with_overflow();
+    if offset_overflow {
+        return &[];
+    }
+    let offset = offset as usize;
+    let len = len as usize;
+    let (end, end_overflow) = offset.overflowing_add(len);
+    if end_overflow || offset >= data.len() {
+        &[]
+    } else {
+        &data[offset..min(end, data.len())]
+    }
+}
+
+#[inline(always)]
+pub fn zero_slice(data: &mut [u8]) {
+    for byte in data {
+        *byte = 0;
+    }
+}
+
+#[inline(always)]
+pub fn copy_slice_padded(
+    src: &[u8],
+    dest: &mut [u8],
+    gas_left: &mut u64,
+) -> Result<(), StatusCode> {
+    consume_copy_cost(gas_left, dest.len() as u64)?;
+    dest[..src.len()].copy_from_slice(src);
+    zero_slice(&mut dest[src.len()..]);
+    Ok(())
+}
+
+#[inline(always)]
+pub fn word_size(bytes: u64) -> Result<u64, StatusCode> {
+    let (end, overflow) = bytes.overflowing_add(31);
+    if overflow {
+        return Err(StatusCode::EVMC_OUT_OF_GAS);
+    }
+    Ok(end / 32)
+}
+
+#[inline(always)]
+pub(super) fn check_min_revision(
+    min_revision: Revision,
+    revision: Revision,
+) -> Result<(), StatusCode> {
+    if revision < min_revision {
+        return Err(StatusCode::EVMC_UNDEFINED_INSTRUCTION);
```

Comment:
This error code is kind of strange in this context. We are talking about revisions, and the error is about instructions. 
I cannot judge until I see how the caller side looks like. I presume this is related to checking for version in certain opcodes, where some sugar will reduce boilerplate. 
Nevertheless, out of context, this symbol is strange.

Reply:
It is used for instructions which have been added in later revisions to fail if the revision is older than the one where the opcode was added.

Reply:
yes, I understand that. But still. you have a function here comparing revisions in a namespace utils... and then returns a undefined_instruction error out of nowhere. 
Maybe this function should be local to the other file, where it shares more context.

Reply:
In my mind `utils` is more or less a collection of stuff that does clearly belong to any particular other module. But I'm fine with moving it.